win32ss/user/ntuser/kbdlayout.c

```diff
@@ -550,54 +595,37 @@ NtUserGetKeyboardLayoutList(
     ULONG nBuff,
     HKL *pHklBuff)
 {
-    UINT uRet = 0;
-    PKL pKl;
+    UINT ret = 0;
+    PWINSTATION_OBJECT pWinSta;
 
     if (!pHklBuff)
         nBuff = 0;
 
     UserEnterShared();
 
-    if (!gspklBaseLayout)
+    if (nBuff > MAXULONG / sizeof(HKL))
     {
-        UserLeave();
-        return 0;
+        SetLastNtError(ERROR_INVALID_PARAMETER);
+        goto Quit;
     }
-    pKl = gspklBaseLayout;
 
-    if (nBuff == 0)
+    _SEH2_TRY
     {
-        do
-        {
-            uRet++;
-            pKl = pKl->pklNext;
-        } while (pKl != gspklBaseLayout);
+        ProbeForWrite(pHklBuff, nBuff * sizeof(HKL), 1);
+        pWinSta = IntGetProcessWindowStation(NULL);
```

Comment:
Why not just calling `IntValidateWindowStationHandle` with the correct parameters?

Reply:
The `IntGetProcessWindowStation` doesn't need to be in the SEH block.
However, the `IntGetKeyboardLayoutList` call definitively needs to be in it.

Reply:
>Why not just calling IntValidateWindowStationHandle

Windows does like this.

Reply:
> The IntGetProcessWindowStation doesn't need to be in the SEH block.
> However, the IntGetKeyboardLayoutList call definitively needs to be in it.

Fixed.

Reply:
> > Why not just calling IntValidateWindowStationHandle
> 
> Windows does like this. Adding `IntGetProcessWindowStation` can simplify the high-level control.

It's not because Windows does something that you have to copy it. Especially if the thing in question is used in only one place.

Reply:
That is used in multiple places. I confirmed.

Reply:
And what about in *our* code?

Reply:
That will be our future.